### PR TITLE
[Tabular] added roc_auc one-vs-one for multi-class AUC classification

### PR DESCRIPTION
**Pull request by gradientsky**
*Description of changes:*
- added multi-class one-vs-one variant of roc_auc

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: VERSION.minor [hunk 1]
```diff
@@ -0,0 +1 @@
+b20210108
```
Comment (Innixma) [inline VERSION.minor:1]: This should be removed

File: core/src/autogluon/core/metrics/__init__.py [hunk 2]
```diff
@@ -354,6 +354,14 @@ def rmse_func(predictions, targets):
                       greater_is_better=True,
                       needs_threshold=True)
 
+roc_auc_ovo_macro = make_scorer('roc_auc_ovo_macro',
```
Comment (Innixma) [inline core/src/autogluon/core/metrics/__init__.py:357]: did you test this with all models in default model set?

File: core/src/autogluon/core/metrics/__init__.py [hunk 3]
```diff
@@ -354,6 +354,14 @@ def rmse_func(predictions, targets):
...
+                                sklearn.metrics.roc_auc_score,
+                                multi_class='ovo',
+                                average='macro',
```
Comment (Innixma) [inline core/src/autogluon/core/metrics/__init__.py:360]: Is macro the only variant that matters?

File: VERSION.minor (hunk 1 above)
Comment (gradientsky) [reply VERSION.minor]: Good catch: removed, added to .gitignore to prevent this happening in future

File: core/src/autogluon/core/metrics/__init__.py (hunk 2 above)
Comment (gradientsky) [reply core/src/autogluon/core/metrics/__init__.py:357]: Yes, I did a few runs on multi-class classification. Catboost doesn't have this metric, I set it to fall back to `AUC:type=Mu`, while FastAI falls back to log_loss. 
```
No output_directory specified. Models will be saved in: AutogluonModels/ag-20210112_013632/
Beginning AutoGluon training ...
AutoGluon will save models to "AutogluonModels/ag-20210112_013632/"
AutoGluon Version:  0.0.16b20210108
Train Data Rows:    891
Train Data Columns: 11
Preprocessing data ...
AutoGluon infers your prediction problem is: 'multiclass' (because dtype of label-column == int, but few unique label-values observed).
	3 unique label values:  [3, 1, 2]
	If 'multiclass' is not the correct problem_type, please manually specify the problem_type argument in fit() (You may specify problem_type as one of: ['binary', 'multiclass', 'regression'])
Train Data Class Count: 3
Using Feature Generators to preprocess the data ...
Fitting AutoMLPipelineFeatureGenerator...
	Available Memory:                    28726.26 MB
	Train Data (Original)  Memory Usage: 0.31 MB (0.0% of available memory)
	Inferring data type of each feature based on column values. Set feature_metadata_in to manually specify special dtypes of the features.
	Stage 1 Generators:
		Fitting AsTypeFeatureGenerator...
	Stage 2 Generators:
		Fitting FillNaFeatureGenerator...
	Stage 3 Generators:
		Fitting IdentityFeatureGenerator...
		Fitting CategoryFeatureGenerator...
			Fitting CategoryMemoryMinimizeFeatureGenerator...
		Fitting TextSpecialFeatureGenerator...
			Fitting BinnedFeatureGenerator...
			Fitting DropDuplicatesFeatureGenerator...
		Fitting TextNgramFeatureGenerator...
			Fitting CountVectorizer for text features: ['Name']
			CountVectorizer fit with vocabulary size = 8
	Stage 4 Generators:
		Fitting DropUniqueFeatureGenerator...
	Types of features in original data (raw dtype, special dtypes):
		('bool', [])         : 1 | ['is_adult']
		('float', [])        : 2 | ['Age', 'Fare']
		('int', [])          : 3 | ['PassengerId', 'SibSp', 'Parch']
		('object', [])       : 4 | ['Sex', 'Ticket', 'Cabin', 'Embarked']
		('object', ['text']) : 1 | ['Name']
	Types of features in processed data (raw dtype, special dtypes):
		('category', [])                    :  5 | ['Sex', 'Ticket', 'Cabin', 'Embarked', 'is_adult']
		('float', [])                       :  2 | ['Age', 'Fare']
		('int', [])                         :  3 | ['PassengerId', 'SibSp', 'Parch']
		('int', ['binned', 'text_special']) : 12 | ['Name.char_count', 'Name.word_count', 'Name.capital_ratio', 'Name.lower_ratio', 'Name.special_ratio', ...]
		('int', ['text_ngram'])             :  9 | ['__nlp__.henry', '__nlp__.john', '__nlp__.master', '__nlp__.miss', '__nlp__.mr', ...]
	0.8s = Fit runtime
	11 features in original data used to generate 31 features in processed data.
	Train Data (Processed) Memory Usage: 0.07 MB (0.0% of available memory)
Data preprocessing and feature engineering runtime = 0.8s ...
AutoGluon will gauge predictive performance using evaluation metric: 'roc_auc_ovo_macro'
	This metric expects predicted probabilities rather than predicted class labels, so you'll need to use predict_proba() instead of predict()
	To change this, specify the eval_metric argument of fit()
Metric roc_auc_ovo_macro is not supported by this model - using AUC:type=Mu instead
Fitting model: NeuralNetMXNet ...
	0.9852	 = Validation roc_auc_ovo_macro score
	9.86s	 = Training runtime
	0.04s	 = Validation runtime
Fitting model: NeuralNetFastAI ...
Metric roc_auc_ovo_macro is not supported by this model - using log_loss instead

█

	0.898	 = Validation roc_auc_ovo_macro score
	9.96s	 = Training runtime
	0.18s	 = Validation runtime
Fitting model: KNeighborsUnif ...
	0.7567	 = Validation roc_auc_ovo_macro score
	0.0s	 = Training runtime
	0.12s	 = Validation runtime
Fitting model: KNeighborsDist ...
	0.7776	 = Validation roc_auc_ovo_macro score
	0.0s	 = Training runtime
	0.11s	 = Validation runtime
Fitting model: RandomForestGini ...
	0.9962	 = Validation roc_auc_ovo_macro score
	0.65s	 = Training runtime
	0.12s	 = Validation runtime
Fitting model: RandomForestEntr ...
	0.9958	 = Validation roc_auc_ovo_macro score
	0.65s	 = Training runtime
	0.11s	 = Validation runtime
Fitting model: ExtraTreesGini ...
	0.9522	 = Validation roc_auc_ovo_macro score
	0.55s	 = Training runtime
	0.12s	 = Validation runtime
Fitting model: ExtraTreesEntr ...
	0.9568	 = Validation roc_auc_ovo_macro score
	0.55s	 = Training runtime
	0.11s	 = Validation runtime
Fitting model: LightGBM ...
	0.998	 = Validation roc_auc_ovo_macro score
	4.58s	 = Training runtime
	0.02s	 = Validation runtime
Fitting model: LightGBMXT ...
	0.9843	 = Validation roc_auc_ovo_macro score
	11.1s	 = Training runtime
	0.02s	 = Validation runtime
Fitting model: CatBoost ...
	0.9988	 = Validation roc_auc_ovo_macro score
	4.09s	 = Training runtime
	0.01s	 = Validation runtime
Fitting model: XGBoost ...
	0.9945	 = Validation roc_auc_ovo_macro score
	1.66s	 = Training runtime
	0.01s	 = Validation runtime
Fitting model: LightGBMCustom ...
	0.9973	 = Validation roc_auc_ovo_macro score
	7.05s	 = Training runtime
	0.02s	 = Validation runtime

█

Fitting model: WeightedEnsemble_L1 ...
	1.0	 = Validation roc_auc_ovo_macro score
	3.79s	 = Training runtime
	0.0s	 = Validation runtime
AutoGluon training complete, total runtime = 57.88s ...

```

File: core/src/autogluon/core/metrics/__init__.py (hunk 3 above)
Comment (gradientsky) [reply core/src/autogluon/core/metrics/__init__.py:360]: There are other options available, but with macro it makes it insensitive to class imbalance.

https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html

**Comment by szha** [comment]
Job PR-859-8 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-859/8/index.html

**Review by Innixma — APPROVED** [review]
